src/psyclone/psyGen.py

```diff
@@ -3083,6 +3083,129 @@ def scalars(self):
+
+            if self._call.vector_index != arg.call.vector_index:
+                # accesses are to different vector indices so do not overlap
+                return False
+        # accesses do overlap
+        return True
+
+    def reset_coverage(self):
+        '''Reset internal state to allow re-use of the object for a different
+        situation.
+
+        '''
+        self._covered = False
+        self._vector_index_access = []
+
+    def update_coverage(self, arg):
+        '''Record any overlap between accesses to the supplied argument and
+        the source argument. If the overlap results in all of the
+        accesses to the source argument being covered (either directly
+        or as a combination with previous arguments) then ensure that
+        the covered() method returns True.
+
+        :param arg: the argument to compare with our internal argument \
+                    and update coverage information
+        :type arg: :py:class:`psyclone.psyGen.Argument`
+
+        '''
+
+        if not self.overlaps(arg):
+            ''' there is no overlap so there is nothing to update '''
+            return
+
+        if isinstance(arg.call, HaloExchange) and \
+           self._arg.vector_size > 1:
+            # The supplied argument is a vector field coming from a
+            # halo exchange and therefore only accesses one of the
+            # vectors
+
+            if isinstance(self._call, HaloExchange):
+                # I am also a halo exchange so only access one of the
+                # vectors. At this point the vector indices of the two
+                # halo exchange fields must be the same, which should
+                # never happen.
+                raise InternalError(
+                    "DataAccess:update_coverage() The halo exchange vector "
+                    "indices for '{0}' are the same. This should never "
+                    "happen".format(self._arg.name))
+            else:
+                # I am not a halo exchange so access all
+                # vectors. However, the supplied argument is a halo
+                # exchange so only accesses one of the vectors. This
+                # results in partial coverage. Therefore record the
+                # index that is accessed and check whether all indices
+                # are now covered
+                if arg.call.vector_index in self._vector_index_access:
+                    raise InternalError(
+                        "DataAccess:update_coverage() Found more than one "
+                        "dependent halo exchange with the same vector index")
+                self._vector_index_access.append(arg.call.vector_index)
+                if len(self._vector_index_access) != self._arg.vector_size:
+                    return
+        # This argument is fully covered
+        self._covered = True
+
+    @property
+    def covered(self):
+        '''Returns true if all of this arguments data has been covered by the
```

Comment:
"the data associated with this argument has been covered..."

Reply:
Modified.